src/Storages/S3Queue/S3QueueFilesMetadata.cpp

```diff
@@ -28,326 +42,875 @@ namespace
+        {
+            return setFileProcessedForOrderedMode(holder);
+        }
+        case S3QueueMode::UNORDERED:
+        {
+            return setFileProcessedForUnorderedMode(holder);
+        }
+    }
 }
 
-String S3QueueFilesMetadata::getMaxProcessedFile()
+void S3QueueFilesMetadata::setFileProcessedForUnorderedMode(ProcessingNodeHolderPtr holder)
 {
-    auto zookeeper = storage->getZooKeeper();
-    return zookeeper->get(zookeeper_processed_path);
+    /// Create a persistent node in /processed and remove ephemeral node from /processing.
+
+    const auto & path = holder->path;
+    const auto node_name = getNodeName(path);
+    const auto node_metadata = createNodeMetadata(path).toString();
+    const auto zk_client = getZooKeeper();
+
+    Coordination::Requests requests;
+    requests.push_back(zkutil::makeCreateRequest(zookeeper_processed_path / node_name, node_metadata, zkutil::CreateMode::Persistent));
+
+    Coordination::Responses responses;
+    if (holder->remove(&requests, &responses))
+    {
+        LOG_TEST(log, "Moved file `{}` to processed", path);
+        return;
+    }
+
+    if (!responses.empty() && responses[0]->error != Coordination::Error::ZOK)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+                        "Cannot create a persistent node in /processed since it already exists");
+    }
+
+    LOG_WARNING(log,
+                "Cannot set file ({}) as processed since ephemeral node in /processing"
+                "does not exist with expected id, "
+                "this could be a result of expired zookeeper session", path);
 }
 
-S3QueueFilesMetadata::S3FilesCollection S3QueueFilesMetadata::getProcessingFiles()
+void S3QueueFilesMetadata::setFileProcessedForOrderedMode(ProcessingNodeHolderPtr holder)
 {
-    auto zookeeper = storage->getZooKeeper();
-    String processing_files;
-    if (!zookeeper->tryGet(zookeeper_processing_path, processing_files))
-        return {};
+    /// Update a persistent node in /processed and remove ephemeral node from /processing.
+
+    const auto & path = holder->path;
+    const auto node_name = getNodeName(path);
+    const auto node_metadata = createNodeMetadata(path).toString();
+    const auto zk_client = getZooKeeper();
+
+    while (true)
+    {
+        std::string res;
+        Coordination::Stat stat;
+        bool exists = zk_client->tryGet(zookeeper_processed_path, res, &stat);
+        Coordination::Requests requests;
+        if (exists)
+        {
+            if (!res.empty())
+            {
+                auto metadata = NodeMetadata::fromString(res);
+                if (metadata.file_path >= path)
+                {
+                    /// Here we get in the case that maximum processed file is bigger than ours.
+                    /// This is possible to achieve in case of parallel processing
+                    /// but for local processing we explicitly disable parallel mode and do everything in a single thread
```

Comment:
Can we also create two S3Queue storages with the same zk path on one instance?

Reply:
Do we have other problems except retries with parallel processing in Ordered mode? 
If only retries, maybe instead of setting `s3queue_processing_threads_num` always to 1, we can set it to 1 only if retries were enabled? Or even throw an exception if retries are enabled in Ordered mode with a message that it can be broken in distributed scenario (and allow to force retries with ordered mode, so we will set threads_num to 1 and won't care about distributed case). Or maybe you already know how to solve this problem and we can just keep it as is and fix later.

Reply:
> Or maybe you already know how to solve this problem and we can just keep it as is and fix later.

Yeah, I have an idea. I think this can be solved if ordered mode was implemented as a combo of ordered and unordered modes. Let's say there is a setting `processing_window`, equal to for, example, 100. This window defines a range of files. Within this window we process files in parallel and do not process any other files until all files within this window are marked as either failed or processed. While we process a window of files we keep track of them in keeper like we track files in unordered mode. But once the full window is processed, we change the information in keeper to contain only max_processed_file (what we aim to do in ordered mode).

> Do we have other problems except retries with parallel processing in Ordered mode?

If server got terminated with SIGABRT and we did not finish processing file `file1`, but another thread already processed `file2`, then on restart we'll think we already processed `file1` (even though we didn't). But this issue can be also solved by my idea described above.

Reply:
> Yeah, I have an idea. ...

Started implementing it. I think it will be better to be put in a separate PR, after this one is merged, for easier review.